embed Button components directly into MenuItem components instead of …

cameronabel opened the pull request:
…nesting as children.

Removes the nested buttons and instead allows the MenuItem component itself act as a button. Ensures consistent behavior when clicking across full width of dropdown menu.
![button_demo](https://github.com/codeforpdx/PASS/assets/56567967/af84bd82-eb9d-4fc4-a10e-ee3b5d7452ea)

cameronabel committed:
embed Button components directly into MenuItem components instead of nesting as children

cameronabel committed:
embed Button components directly into MenuItem components instead of nesting as children

cameronabel committed:
prettier formatting

cameronabel committed:
prettier formatting

xscottxbrownx commented:
What branch did you create this from?  Need to merge in Development into this one, there's lots of updates this is missing.

Seems to work well, love the approach/solution of just making the menu item render as a button component. 

For the color, can you tell if that's a custom material ui color of success? Not sure it's needed but did you test trying to use our theme primary color?  I can't tell if it's the same, it may be.

timbot1789 reviewed (commented):
Thanks for contributing to the project! I like the improvements a lot. Well done.

The code itself looks fine, but could you add some unit tests to the NavMenu as well? If you're not sure how to write them, I'm happy to help with it. Send me a message on discord  and we can set up a call.

@xscottxbrownx unrelated to the PR, but I notice the colors here are hardcoded `rgb` values. Do we have equivalent values in themes we could use here instead?

xscottxbrownx commented:
> @xscottxbrownx unrelated to the PR, but I notice the colors here are hardcoded `rgb` values. Do we have equivalent values in themes we could use here instead?

That was the reason for my last question above your comment.    Honestly, I'm not positive that the colors in the navmenu are our green and red, or if they are the default success and error/danger without testing.  But every time I go to test it seems the solid servers are having issues and I can't login.   Was working fine earlier this morning.

https://github.com/codeforpdx/PASS/assets/71395931/6c4352e4-1ae5-4b1a-afa8-b06be2b11cca

leekahung commented on src/components/NavBar/NavMenu.jsx:
```diff
@@ -52,32 +52,33 @@ const NavMenu = ({ menuId, openMenu, setOpenMenu, anchorEl, setAnchorEl, setShow
+          href={session.info.webId}
+          target="_blank"
+          rel="noreferrer"
+          sx={{ color: 'rgba(1, 121, 105, 1)' }}
```
Think we could use the green color from our existing theme.js

cameronabel replied:
From my (limited) research, it seems that MenuItems do not use the theme palette by default. I have updated theme.js to include this component with the two variants (green/primary and red/error).

leekahung replied:
Think we could have them use the palettes via the `sx` attribute, see clip.

https://github.com/codeforpdx/PASS/assets/14917816/27db94c1-49db-411d-9897-4ae3bc4731cf

Just need to swap what you've made before with the names for the specific color in our palette.

xscottxbrownx replied:
⬆️ This is the solution so we can just change the colors in our theme whenever, and it'll be consistent app wide.

xscottxbrownx replied:
The green is `primary`, the dark red is `secondary`, and that gold/mustard yellow is `tertiary`.   I'm not against leaving logout the `status` color though. Either red works for me.

leekahung commented on src/components/NavBar/NavMenu.jsx:
```diff
@@ -52,32 +52,33 @@ const NavMenu = ({ menuId, openMenu, setOpenMenu, anchorEl, setAnchorEl, setShow
+          component={Button}
+          variant="text"
+          startIcon={<SettingsIcon />}
+          sx={{ color: 'rgba(1, 121, 105, 1)' }}
```
Similar to above, I think we could use the green color from our existing theme.js

leekahung commented on src/components/NavBar/NavMenu.jsx:
```diff
@@ -52,32 +52,33 @@ const NavMenu = ({ menuId, openMenu, setOpenMenu, anchorEl, setAnchorEl, setShow
+          component={Button}
+          startIcon={<LogoutIcon />}
+          onClick={() => setShowConfirmation(true)}
+          sx={{ color: 'rgba(211, 47, 47, 1)' }}
```
Think we can use the same red as the one we set in theme.js

leekahung commented on src/components/NavBar/NavMenu.jsx:
```diff
@@ -52,32 +52,33 @@ const NavMenu = ({ menuId, openMenu, setOpenMenu, anchorEl, setAnchorEl, setShow
-          >
-            Logout
-          </Button>
+        <MenuItem
```
Think we could also get this `<MenuItem>` to be fullWidth like with "Settings" and "Profile". Makes it look nice and consistent.

cameronabel replied:
I believe all three MenuItems are full width. Let me know if I have misunderstood!

leekahung replied:
Similar to what I've seen in the opening post, this is what I saw when testing. 

https://github.com/codeforpdx/PASS/assets/14917816/9db90016-15c1-4fa1-9fca-e86024ceb745

"Setting" and "Profile" looks pretty good having stretched out to the full width of the container, but not for "Log out".

Not really sure what MUI is doing with this component since it does look like it supposed to stretch out, but think we could patch this up by setting `fullWidth` for this `<MenuItem>` like what is done with "Setting".

https://github.com/codeforpdx/PASS/assets/14917816/e144982e-6c7c-415b-be66-955cbf099fbb

cameronabel replied:
Aha, I see what you mean now. I was watching the left side of the button and just assumed it was symmetrical. Will take another look at the logout button this evening.

leekahung reviewed (commented):
Hey @cameronabel, thanks for your new changes to the codebase. It looks pretty good for the most part, so I'm looking forward to seeing the new changes in development.

Just a few comments that I think could be addressed. See feedback.

cameronabel committed:
replace rgba values with theme colors

cameronabel committed:
replace rgba values with theme colors

cameronabel committed:
prettier formatting

cameronabel commented:
> What branch did you create this from? Need to merge in Development into this one, there's lots of updates this is missing.

Good call! I used the link on the issue page, but it seems that branched off master. The newest commit(s) include a rebase onto Development.

> Seems to work well, love the approach/solution of just making the menu item render as a button component.
> 
> For the color, can you tell if that's a custom material ui color of success? Not sure it's needed but did you test trying to use our theme primary color? I can't tell if it's the same, it may be.

From the research I did, it seems that Mui uses a separate styling system for menus that overrides the typography palette, meaning you are unable to just use `color: 'primary'` in the way that buttons can. While I did confirm both rgba values match the theme, this could be a pain to update in the future. ~~I have replaced the rgba values with an import of `theme.js` and reference the values therein.~~ This method didn't sit well, so I have added MenuItem variants to the theme with colors to match the previously used 'primary' and 'error' values.

cameronabel commented:
> Thanks for contributing to the project! I like the improvements a lot. Well done.
> 
> The code itself looks fine, but could you add some unit tests to the NavMenu as well? If you're not sure how to write them, I'm happy to help with it. Send me a message on discord and we can set up a call.
> 
> @xscottxbrownx unrelated to the PR, but I notice the colors here are hardcoded `rgb` values. Do we have equivalent values in themes we could use here instead?

Thank you for your feedback, Tim! I'm not sure what a unit test for this component would look like (some experience writing tests for back-end functionality but limited front-end testing). I'll reach out on Discord.

cameronabel committed:
replace imported theme with variants

leekahung commented on src/components/NavBar/NavMenu.jsx:
```diff
@@ -52,32 +52,27 @@ const NavMenu = ({ menuId, openMenu, setOpenMenu, anchorEl, setAnchorEl, setShow
             Notifications
```
Just noticed this when I have shrunk the screen for the app, see clip below. (Also, ignore the jumping `<MenuList>` for now, haha.)

https://github.com/codeforpdx/PASS/assets/14917816/f8a76544-9610-4d1c-8f0b-ef2cd520cb64

Believe we could do the same modifications we've done for "Profile", "Setting", and "Log out" in this PR and extend it to Notifications.

Looks like we just need to change the base component for the first `<MenuItem>` into a button via `component={Button}`

cameronabel replied:
Oh wow - I hadnt noticed the Notifications component jumping into the menu upon screen resize! I'll take a look at that this evening to see if I can match the appearance to the other MenuItems.

xscottxbrownx replied:
@leekahung @cameronabel 

that was my doing.  We discussed it way back in PR #141  - Check the video there.

I'd have to look at Figma file if we have something different now, but basically on mobile screens all items move into a hamburger menu as is convention. 

I believe the only reason it looks weird right now is because we only have 2 icons on right side of navbar, but honestly the `Messages` route should be a mail icon over by notifications bell icon as well.  So all 3 would go into the hamburger menu on small screens.

I'm still not sure we need the dropdown menu for the routes (`Clients`/`Documents`) - I feel that can just always stay as tabs.  But again, depends on what we add in the future. 

---

Just need to make the `<MenuItem>` a button like you did the rest for consistency.

leekahung replied:
For me, I'm just pointing out the styling for "Notifications". Think we could deal with other issues related to this in another PR.

xscottxbrownx replied:
### Oh gotcha! I misunderstood...

andycwilliams replied:
A little tangential but I've noticed that, when opened, Menu and Notifications use a light green backdrop. This is unlike the `AddClientModal.jsx` and `LogoutModal.jsx` (which have a gray backdrop) and `NewMessage.jsx` (which has a custom blur effect).

`NewMessage.jsx` will be refactored with MUI, so that ought to resolve itself. But in general, I think we should decide on a set style for all modals, pop-ups, etc. Such as whether to stick with gray, green, or something else.

Personally, I think the gray backdrop is a bit nicer as well as more standard-looking. The green is a tad distracting in my opinion, even though it does fit in with our current color scheme (except we may not even end up sticking with that). Also, the gray is what comes default with MUI modals.

xscottxbrownx replied:
> A little tangential but I've noticed that, when opened, Menu and Notifications use a light green backdrop. This is unlike the `AddClientModal.jsx` and `LogoutModal.jsx` (which have a gray backdrop) and `NewMessage.jsx` (which has a custom blur effect).
> 
> `NewMessage.jsx` will be refactored with MUI, so that ought to resolve itself. But in general, I think we should decide on a set style for all modals, pop-ups, etc. Such as whether to stick with gray, green, or something else.
> 
> Personally, I think the gray backdrop is a bit nicer as well as more standard-looking. The green is a tad distracting in my opinion, even though it does fit in with our current color scheme (except we may not even end up sticking with that). Also, the gray is what comes default with MUI modals.

Yeah, something I did keeping it all "on theme".  But agree, kind of looks a little tacky.   

And yeah we'd have to deep dive into the `theme.js` if we wanted it all to be the green as styling the modal backdrops with sx didn't seem to work no matter what I tried.

But either direction, again consistency. All grey or all green.  But seeing them both, I think the grey looks better honestly.

timbot1789 replied:
+1 for having the gray backdrop for modals.
-1 for Scott spelling gray like a british person. This Is America.

xscottxbrownx replied:
> +1 for having the gray backdrop for modals.
> -1 for Scott spelling gray like a british person. This Is America.

“Gray” is a last name, “grey” is a color - despite the last decade and a half of Americans changing it because they don’t know the difference 😂

andycwilliams commented:
I've noticed the same issue in `NavbarLinks.jsx` when the tabs are condensed into a menu for mobile views. I don't think it would be too much for this PR to implement the same fix there.

xscottxbrownx commented:
@cameronabel are these quick fixes something you are still interested in / have time for?

cameronabel commented:
> @cameronabel are these quick fixes something you are still interested in / have time for?

@xscottxbrownx 
Yep! My day job is just very "end-of-the-month-quota-oriented" and I have been slammed hitting June's numbers. Will have some more time over the next few days.

cameronabel committed:
full-width menu items under hamburger menu

cameronabel committed:
full-width menu items under navigation menu

cameronabel committed:
prettier and revert changes to prettier script'

leekahung reviewed (approved):
Just saw the recent changes to the PR that touched on some of the points I've brought up before. Looking pretty good as is. Think we just need to handle the merge conflict. I'll approve this after that gets cleared up

---

Update: Messed up with the approval. I mean to use comment.

cameronabel commented:
This has been a learning experience with MUI for sure!

![chrome_k4HN9z6wg5](https://github.com/codeforpdx/PASS/assets/56567967/f258d8c6-3b42-464b-9725-a4d4daba5986)

I think the behavior is now as expected for both the hamburger menu and the Nav menu items. Colors should also all be assigned via the theme instead of matching color codes.

I noticed some odd behavior when the menu collapsed - the menu items beneath the horizontal divider would both shrink to become even more narrow than the logout button in Ka Hung's post above. Namely: the Settings menu item would become narrow when the menu was collapsed, but full width when the menu was expanded. Is this set up in a template or style somewhere? Regardless - for consistency, I set all of the menu items explicitly to full width (even those that were not shrinking on menu resize).

Edit: An unintended consequence of the move from nested buttons is a "thinning" of the selection area for each button/menu item. The actual selectable area for each button is technically just as thick, but we've lost some of the button padding which makes the menu feel more cramped (particularly for touchscreen users):
![image](https://github.com/codeforpdx/PASS/assets/56567967/3bf0fae4-7890-45d6-b094-a76ec4ffecf5)

timbot1789 reviewed (commented):
Code itself looks good to me. Thanks for the contribution!

To fix the linter errors, you can run `npm run lint:fix & npm run prettier:run` to clean up a lot of these issues for you. Anything the computer can't fix itself, it will tell you about.

cameronabel committed:
merging updates since last branch

cameronabel commented:
> Code itself looks good to me. Thanks for the contribution!
> 
> To fix the linter errors, you can run `npm run lint:fix & npm run prettier:run` to clean up a lot of these issues for you. Anything the computer can't fix itself, it will tell you about.

@timbot1789 I'm on it :) looks like there was an update to Development since I branched this initially that clashes - making a few updates and recommitting shortly...

cameronabel committed:
updated changes to merge with updates to Dev branch

cameronabel commented:
> Just saw the recent changes to the PR that touched on some of the points I've brought up before. Looking pretty good as is. Think we just need to handle the merge conflict. I'll approve this after that gets cleared up
> 
> Update: Messed up with the approval. I mean to use comment.

@leekahung Conflicts should be resolved now - would you mind a final review and approval? Thanks!

leekahung reviewed (approved):
I've just checked out the code and build on my local machine. Think what I've touched on has been addressed. I'm good approving this.
